Error in benchmarks

```
>       a = APyCFixed.from_complex(
            np.random.rand(1)[0] - 0.5 + 1j * (np.random.rand(1)[0] - 0.5),
            bits=20,
            int_bits=1,
        )
E       ValueError: Non-supported type: <class 'numpy.complex128'>
```

One solution is to convert to standard Python complex, but it would be nice if the numpy scalar types can be supported.

--- a/lib/test/apycfixed/test_constructors.py
+++ b/lib/test/apycfixed/test_constructors.py
@@ -26,3 +26,34 @@ def test_raises():
 def test_constructor():
     assert APyCFixed((0x14,), int_bits=4, frac_bits=4) == 1.25
     assert APyCFixed((0x00, 0x14), int_bits=4, frac_bits=4) == 0.0 + 1.25j
+
+
+def test_from_float():
+    np = pytest.importorskip("numpy")
+    types = [
+        np.float16,
+        np.float32,
+        np.float64,
+        np.int8,
+        np.int16,
+        np.int32,
+        np.int64,
+        np.uint8,
+        np.uint16,
+        np.uint32,
+        np.uint64,
+    ]
+    for t in types:
+        a = APyCFixed.from_float(t(125), int_bits=30, frac_bits=30)
+        assert a.is_identical(APyCFixed.from_complex(125, int_bits=30, frac_bits=30))
+
+
+def test_from_complex():
+    np = pytest.importorskip("numpy")
+    types = [
+        np.complex64,
+        np.complex128,
+    ]
+    for t in types:
+        a = APyCFixed.from_float(t(923.75), int_bits=30, frac_bits=30)
+        assert a.is_identical(APyCFixed.from_complex(923.75, int_bits=30, frac_bits=30))

--- a/lib/test/apycfloat/test_constructors.py
+++ b/lib/test/apycfloat/test_constructors.py
@@ -92,3 +92,32 @@ def test_from_complex():
     assert APyCFloat.from_complex(
         APyCFixed.from_complex(-4, int_bits=10, frac_bits=8), exp_bits=10, man_bits=8
     ).is_identical(APyCFloat((True, 0), (2**9 + 1, 0), (0, 0), exp_bits=10, man_bits=8))
+
+    np = pytest.importorskip("numpy")
+    types = [
+        np.complex64,
+        np.complex128,
+    ]
+    for t in types:
+        a = APyCFloat.from_float(t(923.75), exp_bits=11, man_bits=52)
+        assert a.is_identical(APyCFloat.from_complex(923.75, exp_bits=11, man_bits=52))
+
+
+def test_from_float():
+    np = pytest.importorskip("numpy")
+    types = [
+        np.float16,
+        np.float32,
+        np.float64,
+        np.int8,
+        np.int16,
+        np.int32,
+        np.int64,
+        np.uint8,
+        np.uint16,
+        np.uint32,
+        np.uint64,
+    ]
+    for t in types:
+        a = APyCFloat.from_float(t(125), exp_bits=11, man_bits=52)
+        assert a.is_identical(APyCFloat.from_complex(125, exp_bits=11, man_bits=52))

--- a/src/apycfixed.cc
+++ b/src/apycfixed.cc
@@ -668,22 +668,21 @@ APyCFixed APyCFixed::from_number(
 )
 {
     if (nb::isinstance<nb::int_>(py_obj)) {
-        return APyCFixed::from_integer(
-            nb::cast<nb::int_>(py_obj), int_bits, frac_bits, bits
-        );
+        const nb::int_& val = nb::cast<nb::int_>(py_obj);
+        return APyCFixed::from_integer(val, int_bits, frac_bits, bits);
     } else if (nb::isinstance<nb::float_>(py_obj)) {
-        const auto d = static_cast<double>(nb::cast<nb::float_>(py_obj));
-        return APyCFixed::from_double(d, int_bits, frac_bits, bits);
+        double val = static_cast<double>(nb::cast<nb::float_>(py_obj));
+        return APyCFixed::from_double(val, int_bits, frac_bits, bits);
     } else if (nb::isinstance<APyFixed>(py_obj)) {
+        const APyFixed& fx = nb::cast<APyFixed>(py_obj);
         APyCFixed result(int_bits, frac_bits, bits);
-        const auto d = static_cast<APyFixed>(nb::cast<APyFixed>(py_obj));
         fixed_point_cast_unsafe(
-            std::begin(d._data),
-            std::end(d._data),
+            std::begin(fx._data),
+            std::end(fx._data),
             result.real_begin(),
             result.real_end(),
-            d._bits,
-            d._int_bits,
+            fx._bits,
+            fx._int_bits,
             result._bits,
             result._int_bits,
             QuantizationMode::RND_INF,
@@ -691,15 +690,15 @@ APyCFixed APyCFixed::from_number(
         );
         return result;
     } else if (nb::isinstance<APyFloat>(py_obj)) {
+        const auto fx = nb::cast<APyFloat>(py_obj).to_fixed();
         APyCFixed result(int_bits, frac_bits, bits);
-        const auto d = static_cast<APyFloat>(nb::cast<APyFloat>(py_obj)).to_fixed();
         fixed_point_cast_unsafe(
-            std::begin(d._data),
-            std::end(d._data),
+            std::begin(fx._data),
+            std::end(fx._data),
             result.real_begin(),
             result.real_end(),
-            d._bits,
-            d._int_bits,
+            fx._bits,
+            fx._int_bits,
             result._bits,
             result._int_bits,
             QuantizationMode::RND_INF,
@@ -707,19 +706,27 @@ APyCFixed APyCFixed::from_number(
         );
         return result;
     } else if (nb::isinstance<APyCFixed>(py_obj)) { // One should really use cast
-        const auto d = static_cast<APyCFixed>(nb::cast<APyCFixed>(py_obj));
-        return d.cast(
+        const APyCFixed& fx = nb::cast<APyCFixed>(py_obj);
+        return fx.cast(
             int_bits, frac_bits, QuantizationMode::RND_INF, OverflowMode::WRAP, bits
         );
     } else if (nb::isinstance<std::complex<double>>(py_obj)) {
-        const auto c = nb::cast<std::complex<double>>(py_obj);
-        return APyCFixed::from_complex(c, int_bits, frac_bits, bits);
+        std::complex<double> cplx = nb::cast<std::complex<double>>(py_obj);
+        return APyCFixed::from_complex(cplx, int_bits, frac_bits, bits);
     } else {
-        const nb::type_object type = nb::cast<nb::type_object>(py_obj.type());
-        const nb::str type_string = nb::str(type);
-        throw std::domain_error(
-            std::string("Non-supported type: ") + type_string.c_str()
-        );
+        // Last resort, try casting the Python object to a `std::complex`. This is
+        // useful since the Python type `numpy.complex128` does not match any nanobind
+        // type.
+        std::complex<double> cplx;
+        if (nb::try_cast<std::complex<double>>(py_obj, cplx)) {
+            return APyCFixed::from_complex(cplx, int_bits, frac_bits, bits);
+        } else {
+            const nb::type_object type = nb::cast<nb::type_object>(py_obj.type());
+            const nb::str type_string = nb::str(type);
+            throw std::domain_error(
+                std::string("Non-supported type: ") + type_string.c_str()
+            );
+        }
     }
 }
 

--- a/src/apycfloat.cc
+++ b/src/apycfloat.cc
@@ -168,27 +168,24 @@ APyCFloat APyCFloat::from_number(
     const nb::object& py_obj, int exp_bits, int man_bits, std::optional<exp_t> bias
 )
 {
+    exp_t res_bias = bias.value_or(ieee_bias(exp_bits));
     if (nb::isinstance<nb::int_>(py_obj)) {
-        const nb::int_& py_int = nb::cast<nb::int_>(py_obj);
-        return APyCFloat::from_integer(py_int, exp_bits, man_bits, bias);
+        const nb::int_& val = nb::cast<nb::int_>(py_obj);
+        return APyCFloat::from_integer(val, exp_bits, man_bits, res_bias);
     } else if (nb::isinstance<nb::float_>(py_obj)) {
-        double d = static_cast<double>(nb::cast<nb::float_>(py_obj));
-        return APyCFloat::from_double(d, exp_bits, man_bits, bias);
+        double val = static_cast<double>(nb::cast<nb::float_>(py_obj));
+        return APyCFloat::from_double(val, exp_bits, man_bits, res_bias);
     } else if (nb::isinstance<std::complex<double>>(py_obj)) {
-        std::complex<double> c = nb::cast<std::complex<double>>(py_obj);
-        return APyCFloat::from_complex(c, exp_bits, man_bits, bias);
+        std::complex<double> cplx = nb::cast<std::complex<double>>(py_obj);
+        return APyCFloat::from_complex(cplx, exp_bits, man_bits, res_bias);
     } else if (nb::isinstance<APyFloat>(py_obj)) {
-        exp_t res_bias = bias.value_or(ieee_bias(exp_bits));
-        APyFloatSpec res_spec { std::uint8_t(exp_bits),
-                                std::uint8_t(man_bits),
-                                res_bias };
-        APyFloat c = nb::cast<APyFloat>(py_obj);
-        if (res_spec != c.spec()) {
-            c = c.cast(exp_bits, man_bits, res_bias, QuantizationMode::RND_CONV);
+        APyFloatSpec spec { std::uint8_t(exp_bits), std::uint8_t(man_bits), res_bias };
+        APyFloat fp = nb::cast<APyFloat>(py_obj);
+        if (spec != fp.spec()) {
+            fp = fp.cast(exp_bits, man_bits, res_bias, QuantizationMode::RND_CONV);
         }
-        return APyCFloat(c.get_data(), exp_bits, man_bits, res_bias);
+        return APyCFloat(fp.get_data(), exp_bits, man_bits, res_bias);
     } else if (nb::isinstance<APyFixed>(py_obj)) {
-        exp_t res_bias = bias.value_or(ieee_bias(exp_bits));
         const APyFixed& fx = nb::cast<const APyFixed&>(py_obj);
         const APyFloatData& data = floating_point_from_fixed_point(
             std::begin(fx._data),
@@ -201,7 +198,6 @@ APyCFloat APyCFloat::from_number(
         );
         return APyCFloat(data, exp_bits, man_bits, res_bias);
     } else if (nb::isinstance<APyCFixed>(py_obj)) {
-        exp_t res_bias = bias.value_or(ieee_bias(exp_bits));
         const APyCFixed& fx = nb::cast<const APyCFixed&>(py_obj);
         const APyFloatData& re_data = floating_point_from_fixed_point(
             fx.real_cbegin(),
@@ -223,11 +219,19 @@ APyCFloat APyCFloat::from_number(
         );
         return APyCFloat(re_data, im_data, exp_bits, man_bits, res_bias);
     } else {
-        const nb::type_object type = nb::cast<nb::type_object>(py_obj.type());
-        const nb::str type_string = nb::str(type);
-        throw std::domain_error(
-            std::string("Non supported type: ") + type_string.c_str()
-        );
+        // Last resort, try casting the Python object to a `std::complex`. This is
+        // useful since the Python type `numpy.complex128` does not match any nanobind
+        // type.
+        std::complex<double> cplx;
+        if (nb::try_cast<std::complex<double>>(py_obj, cplx)) {
+            return APyCFloat::from_complex(cplx, exp_bits, man_bits, res_bias);
+        } else {
+            const nb::type_object type = nb::cast<nb::type_object>(py_obj.type());
+            const nb::str type_string = nb::str(type);
+            throw std::domain_error(
+                std::string("Non supported type: ") + type_string.c_str()
+            );
+        }
     }
 }
 

--- a/src/apyfixedarray.cc
+++ b/src/apyfixedarray.cc
@@ -1363,16 +1363,15 @@ APyFixedArray APyFixedArray::from_numbers(
 
     // Extract all Python doubles and integers
     auto py_objs = python_sequence_walk<nb::float_, nb::int_, APyFixed, APyFloat>(
-        number_seq, "APyFixedArray.from_complex"
+        number_seq, "APyFixedArray.from_float"
     );
 
     // Iterate over objects and set in fixed-point array
     for (std::size_t i = 0; i < result._nitems; i++) {
         if (nb::isinstance<nb::float_>(py_objs[i])) {
             // Python double object
-            double d = static_cast<double>(nb::cast<nb::float_>(py_objs[i]));
             fixed_point_from_double(
-                d,
+                static_cast<double>(nb::cast<nb::float_>(py_objs[i])),
                 std::begin(result._data) + (i + 0) * result._itemsize,
                 std::begin(result._data) + (i + 1) * result._itemsize,
                 result._bits,

--- a/src/python_util.h
+++ b/src/python_util.h
@@ -501,38 +501,38 @@ python_sequence_walk(const nanobind::sequence& py_seq, std::string_view err_pref
 {
     namespace nb = nanobind;
 
-    // Result output iterator
-    std::vector<nb::object> result {};
-
     // Walk the Python sequences and extract the data
+    using iterator_t = decltype(std::begin(py_seq));
     struct seq_it_pair {
-        decltype(std::begin(py_seq)) iterator;
-        decltype(std::end(py_seq)) sentinel;
+        iterator_t iterator;
+        iterator_t sentinel;
     };
     std::stack<seq_it_pair, std::vector<seq_it_pair>> it_stack;
     it_stack.push({ py_seq.begin(), py_seq.end() });
 
+    std::vector<nb::object> result {};
     while (!it_stack.empty()) {
-        if (it_stack.top().iterator == it_stack.top().sentinel) {
+        auto&& iterator = it_stack.top().iterator;
+        auto&& sentinel = it_stack.top().sentinel;
+        if (iterator == sentinel) {
             // End of current iterator/sentinel pair. Pop it.
             it_stack.pop();
         } else {
-            if (is_seq_and_exclude<nb::str, PyTypes...>(*it_stack.top().iterator)) {
+            if (is_seq_and_exclude<nb::str, PyTypes...>(*iterator)) {
                 // New sequence found. We need to go deeper
-                auto new_sequence = nb::cast<nb::sequence>(*it_stack.top().iterator++);
-                it_stack.push({ new_sequence.begin(), new_sequence.end() });
-            } else if ((nb::isinstance<PyTypes>(*it_stack.top().iterator) || ...)) {
+                auto&& new_sequence = nb::cast<nb::sequence>(*iterator++);
+                it_stack.push({ std::begin(new_sequence), std::end(new_sequence) });
+            } else if ((nb::isinstance<PyTypes>(*iterator) || ...)) {
                 // Element matching one of the PyTypes found, store it in container
-                result.emplace_back(nb::cast<nb::object>(*it_stack.top().iterator++));
+                result.emplace_back(nb::cast<nb::object>(*iterator++));
             } else {
-                nb::object err_obj = nb::cast<nb::object>(*it_stack.top().iterator);
-                throw std::domain_error(
-                    fmt::format(
-                        "{}: unexpected type when traversing Sequence: {}",
-                        err_prefix,
-                        nb::repr(nb::cast<nb::type_object>(err_obj.type())).c_str()
-                    )
+                auto&& err_obj = nb::cast<nb::object>(*iterator);
+                std::string err_msg = fmt::format(
+                    "{}: unexpected type when traversing Sequence: {}",
+                    err_prefix,
+                    nb::repr(nb::cast<nb::type_object>(err_obj.type())).c_str()
                 );
+                throw std::domain_error(err_msg);
             }
         }
     }

